pubtools/exodus/hooks/pulp.py

```diff
@@ -0,0 +1,165 @@
+                    [str(role) for role in typed_ctx["roles"]],
+                )
+                break
+        else:
+            LOG.debug("You are not authenticated with exodus-gw.")
+
+    @classmethod
+    def _unpack_response(cls, response):
+        """Raise if 'response' was not successful.
+
+        This is the same as response.raise_for_status(), merely wrapping it
+        to ensure the body is logged when possible."""
+
+        try:
+            response.raise_for_status()
+        except Exception as outer:
+            try:
+                body = response.json()
+            except:
+                raise outer
+
+            LOG.error("Unsuccessful response from exodus-gw: %s", body)
+            raise
+
+    @backoff.on_predicate(
+        wait_gen=backoff.expo,
+        predicate=lambda task: task["state"] not in ("FAILED", "COMPLETE"),
+        max_time=1800,
+    )
+    def _poll_commit_completion(self, commit):
+        """Issues request(s) to exodus-gw to get the task state, returning once
+        it either is FAILED or COMPLETED."""
+        task_url = urljoin(self._gw_url, commit["links"]["self"])
+
+        resp = self._do_request(method="GET", url=task_url)
+        task = resp.json()
+
+        LOG.info("Task state: %s", task["state"])
+
+        return task
+
+    @hookimpl
+    def pulp_repository_pre_publish(self, to_publish):
+        """Invoked before a Pulp repository is published.
+
+        This implementation adds to each config the --exodus-publish argument,
+        attaching the repository to an Exodus gateway publish.
+
+        Args:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of Distributor-config pairs intended for publishing.
+
+        Returns:
+            to_publish (list[tuple[:class:`~pubtools.pulplib.Distributor`, dict]])
+                List of potentially adjusted Distributor-config pairs to publish.
+        """
+
+        if not self._publish:
+            publish_url = os.path.join(self._gw_url, self._gw_env, "publish")
+            resp = self._do_request(method="POST", url=publish_url)
+            self._publish = resp.json()
+            LOG.debug("Created publish: %s", self._publish["id"])
+
+        publish_id = self._publish["id"]
+
+        for (_, config) in to_publish:
+            rsync_extra_args = config.get("rsync_extra_args", [])
+            rsync_extra_args.append("--exodus-publish=%s" % publish_id)
+            config["rsync_extra_args"] = rsync_extra_args
+
+        return to_publish
+
+    @hookimpl
+    def task_pulp_flush(self):
+        """Invoked during task execution after successful completion of all
+        Pulp writes and publishes.
+
+        This implementation commits the active Exodus gateway publish, making
+        the content visible on the target CDN environment.
+        """
+
+        if not self._publish:
+            LOG.info("No publish to commit.")
```

Comment:
Log messages again should mention something about exodus-gw since they will be mixed in with other logs.